libs/langchain-community/src/vectorstores/turbopuffer.ts

```diff
@@ -0,0 +1,280 @@
+
+  get lc_secrets(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  get lc_aliases(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  // Handle minification for tracing
+  static lc_name(): string {
+    return "TurbopufferVectorStore";
+  }
+
+  protected distanceMetric: TurbopufferDistanceMetric = "cosine_distance";
+
+  protected apiKey: string;
+
+  protected namespace = "default";
+
+  protected apiUrl = "https://api.turbopuffer.com/v1";
+
+  caller: AsyncCaller;
+
+  batchSize = 500;
+
+  public _vectorstoreType(): string {
+    return "turbopuffer";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: TurbopufferParams) {
+    super(embeddings, args);
+
+    const {
+      apiKey: argsApiKey,
+      namespace,
+      distanceMetric,
+      apiUrl,
+      batchSize,
+      ...asyncCallerArgs
+    } = args;
+
+    const apiKey = argsApiKey ?? getEnvironmentVariable("TURBOPUFFER_API_KEY");
+    if (!apiKey) {
+      throw new Error(
+        `Turbopuffer API key not found.\nPlease pass it in as "apiKey" or set it as an environment variable called "TURBOPUFFER_API_KEY"`
+      );
+    }
+    this.apiKey = apiKey;
+    this.namespace = namespace ?? this.namespace;
+    this.distanceMetric = distanceMetric ?? this.distanceMetric;
+    this.apiUrl = apiUrl ?? this.apiUrl;
+    this.batchSize = batchSize ?? this.batchSize;
+    this.caller = new AsyncCaller({
+      maxConcurrency: 6,
+      maxRetries: 0,
+      ...asyncCallerArgs,
+    });
+  }
+
+  getJsonHeader(): TurbopufferHeaders {
+    return {
+      headers: {
+        Authorization: `Bearer ${this.apiKey}`,
+        "Content-Type": "application/json",
+      },
+    };
+  }
+
+  async callWithRetry(fetchUrl: string, stringifiedBody: string) {
+    const response = await this.caller.call(async () =>
+      fetch(fetchUrl, {
+        method: "POST",
+        headers: this.getJsonHeader().headers,
+        body: stringifiedBody,
+      })
+    );
+    return response;
+  }
+
+  async addVectors(
+    vectors: number[][],
+    documents: DocumentInterface[],
+    options?: { ids?: number[] }
+  ): Promise<void> {
+    try {
+      if (options?.ids && options.ids.length !== vectors.length) {
+        throw new Error(
+          "Number of ids provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length !== vectors.length) {
+        throw new Error(
+          "Number of documents provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length === 0) {
+        throw new Error("No documents provided");
+      }
+
+      const batchedVectors = chunkArray(vectors, this.batchSize);
+      const batchedDocuments = chunkArray(documents, this.batchSize);
+      const batchedIds = options?.ids
+        ? chunkArray(options.ids, this.batchSize)
+        : batchedDocuments.map((docs, index) =>
```

Comment:
We definitely don't want to assign ids numerically like this because they'll be overwritten on subsequent insert. Use `uuid` or even just don't specify them if Turbopuffer handles it?

Reply:
Turbopuffer currently only supports 64 bit unsigned integer values. Using `v4` from `uuid` will give 128-bit values. Do you have a preference for how they are truncated @jacoblee93 ?

@sirupsen do you have any idea on timeline for 128 bit values or preference here also?

Reply:
Good to know! No preference, but if we can just let turbopuffer auto-assign, that seems like the best approach.

CC @bracesproul who's been taking over landing this.

Reply:
@jacoblee93 @mattzcarey let me try to see how far I can get on 128 bit / UUID support tomorrow... I'll circle back then. Randomizing a 64 bit integer isn't terrible at these low counts, but still not collision numbers I'm comfortable with given the behaviour on a collision (overriding a previous value).

Reply:
Ok, @bracesproul let's get the rest in good shape and wait another day then 👍

Reply:
Hey... give me till next week, I ran into generic hell in Rust shoving in the string type for ids, before we introduce proper schema support. I'll get it working, just needed more than today :/

Reply:
No problem at all! Just ping me here or in Slack

Reply:
hey @sirupsen have you gotten around to 128 bit yet? No worries if not, just checking back in.

Reply:
Chatted with them externally, still in progress!

Reply:
shipped monday!